volatility3/framework/objects/utility.py

```diff
@@ -94,16 +99,93 @@ def pointer_to_string(
         count=count,
         errors=errors,
         block_size=block_size,
+        encoding=encoding,
     )
 
 
+def gather_contiguous_bytes_from_address(
+    context, data_layer, starting_address: int, count: int
+) -> bytes:
+    """
+    This method reconstructs a string from memory while also carefully examining each page
+
+    It goes page-by-page reading the bytes. This is done by calculating page boundaries
+    and then only reading one page at a time.
+
+    If a page is missing, the code initially catches the exception.
+    If data is non-empty (meaning at least one read succeeded), then we return what was read
+    If the first page fails, then we re-raise the exception
+    """
+
+    data = b""
+
+    last_address = None
+
+    for address, length, _, _, _ in data_layer.mapping(
+        offset=starting_address, length=count, ignore_errors=True
+    ):
+        # Used to track when we hit a paged out page
+        if not last_address:
+            last_address = address + length
+
+        # we hit a swapped out page
+        elif last_address and last_address != address:
+            break
+
+        data += data_layer.read(address, length)
+
+    # if we were able to read from the first page, we want to try and construct the string
+    # if the first page fails -> throw exception
+    if data:
+        return data
+    else:
+        raise exceptions.InvalidAddressException(
+            layer_name=data_layer, invalid_address=starting_address
+        )
+
+    return data
+
+
+def bytes_to_decoded_string(data: bytes, encoding: str, errors: str) -> bytes:
+    """
+    This function takes a bytes buffer that contains at a string of unknown
+    length starting at the first byte, and returns the properly decoded string
+
+    It starts by using Python's `bytes.decode` to attempt to decode the entire string
+    It then finds the termination character (\ufffd or \x00) and splices the string
+    Finally, it returns this spliced string after its been decoded with the
+        caller-specified encoding
+    """
+    # this is the standard byte used to replace bad unicode characters
+    unicode_replacement_char = "\ufffd"
+
+    # used to find the terminating byte
+    termination_re = re.compile(f"{unicode_replacement_char}|\x00")
+
+    # run over the entire string, letting Python replace invalid characters
+    full_decoded_string = data.decode(encoding=encoding, errors="replace")
+
+    # stop at the first terminating character or get the whole string if not found
+    try:
+        idx = termination_re.search(full_decoded_string).start()
+    except AttributeError:
+        idx = len(full_decoded_string)
```

Comment:
If no terminator is found, then we return the entire length of the string, and if we hit a badly encoded character, it'll return everything that was well encoded up until then?  The docstring should definitely document that (ideally in an 
```
Args:
  data: ...
  encoding: ...
  errors: ...
  
Returns:
  ...
```
-style docstring...

We may want it to throw an error if something didn't decode correctly (or the user should have *some* way to know that things didn't go as expected) rather than them complaining of truncated strings and us scratching our heads...

Reply:
I added a new parameter for this to give the caller control and documented it all.